src/Nethermind/Nethermind.Runner/Ethereum/Steps/EthereumStepsManager.cs

```diff
@@ -19,147 +19,98 @@
-                        ?? false;
-
-                    if (contextType == typeof(object))
-                    {
-                        return null;
-                    }
-
-                    Type stepTypeForContext = sameStepSubtypes.Where(t => !t.IsAbstract)
-                        .FirstOrDefault(t => HasConstructorWithParameter(t, contextType));
-
-                    return stepTypeForContext != null
-                        ? stepTypeForContext
-                        : GetStepTypeRecursive(contextType?.BaseType);
-                }
-
-                return sameStepSubtypes.Count() == 1 ? sameStepSubtypes.First() : GetStepTypeRecursive(_context.GetType());
+                throw new ArgumentNullException(nameof(loader));
             }
 
-            foreach (IGrouping<Type?, Type> typeGroup in types.Where(t => t != null))
-            {
-                if (cancellationToken.IsCancellationRequested)
-                {
-                    break;
-                }
-                
-                Type? type = GetStepType(typeGroup);
-                if (type != null)
-                {
-                    if (_logger.IsDebug) _logger.Debug($"Discovered Ethereum step: {type.Name}");
-                    _discoveredSteps[type] = false;
-
-                    Type? baseType = GetStepBaseType(type);
-                    if (baseType != null)
-                    {
-                        _hasFinishedExecution[baseType] = false;
-                    }
-                }
-            }
+            _context = context ?? throw new ArgumentNullException(nameof(context));
+            _logger = logManager?.GetClassLogger<EthereumStepsManager>()
+                      ?? throw new ArgumentNullException(nameof(logManager));
 
-            await ReviewDependencies(cancellationToken);
+            _allSteps = loader.LoadSteps(_context.GetType()).ToList();
+            _allStepsByBaseType = _allSteps.ToDictionary(s => s.StepBaseType, s => s);
         }
 
-        private readonly ConcurrentDictionary<Type, bool> _hasFinishedExecution = new ConcurrentDictionary<Type, bool>();
-
-        private readonly ConcurrentDictionary<Type, bool> _discoveredSteps = new ConcurrentDictionary<Type, bool>();
-
-        private static bool IsStepType(Type? t) => t != null && typeof(IStep).IsAssignableFrom(t);
-
-        private Type? GetStepBaseType(Type? type) => IsStepType(type?.BaseType) ? GetStepBaseType(type?.BaseType) : type;
-
         private async Task ReviewDependencies(CancellationToken cancellationToken)
         {
-            List<Type> typesReady = new List<Type>();
             bool changedAnything;
             do
             {
-                typesReady.Clear();
+                foreach (StepInfo stepInfo in _allSteps)
+                {
+                    _logger.Debug($"{stepInfo} is {stepInfo.Stage}");
+                }
+                
+                await _autoResetEvent.WaitOneAsync(cancellationToken);
```

Comment:
I am quite positive this can hang indefinitely on some scenarios. Wouldn't it be better to just cancel the whole process then? 

What are those scenarios? 
1. Step hangs
2. Step fails? (maybe not)
3. ???

Reply:
If step hangs then the problem is in the step itself (some steps last many hours)
step fails - hen it is market as complete and some other steps may fail because of lacking initialization and it is logged